lrlex/src/lib/parser.rs

```diff
@@ -379,51 +379,85 @@ impl<StorageT: TryFrom<usize>> LexParser<StorageT> {
+                //
+                // If regex_syntax changes behavior here, it is highly likely
+                // that we should just remove this assertion, and send regex `\<`.
+                // Instead of unescaping `\<` into `<`.
+                //
+                // I'm fairly convinced that if this changes we should just remove
+                // this assertion, still it may be worthwhile to ensure that we notice such a change.
+                debug_assert!(regex_syntax::is_meta_character('<').not());
                 let re_str: &str = re.borrow();
                 let mut re_chars = re_str.char_indices().peekable();
-                debug_assert!(regex_syntax::is_meta_character('<').not());
+
+                // Look for an escape sequence which needs unescaping,
+                // If we don't find one, don't bother allocating a string
+                // on which to perform unescaping.  Just return the unmodified parameter.
                 let mut cursor = loop {
                     if let Some((i, c)) = re_chars.next() {
                         if c == '\\' {
-                            if let Some((j, _)) = re_chars.peek() {
+                            // Look at the next character and whether it is something we need to unescape
+                            if let Some((j, peeked_char)) = re_chars.peek() {
                                 let s = &re_str[*j..];
-                                if !(regex_syntax::is_meta_character(s.chars().next().unwrap())
+                                if !(regex_syntax::is_meta_character(*peeked_char)
                                     || RE_LEX_ESC.is_match(s))
                                 {
-                                    break Some((i, s, j));
+                                    break (Some((i, s, j, peeked_char)));
                                 }
                             }
                         }
                     } else {
                         break None;
                     }
                 };
+
                 if cursor.is_none() {
+                    // There is nothing to unescape.
                     return re;
                 }
-                let mut unescaped = re_str.to_owned();
-                let mut diff = 0usize;
-                'outer: loop {
-                    if let Some((i, s, j)) = cursor {
-                        if !(regex_syntax::is_meta_character(s.chars().next().unwrap())
-                            || RE_LEX_ESC.is_match(s))
-                        {
-                            unescaped.replace_range(i - diff..*j - diff, "");
-                            diff += 1;
-                        }
-                    } else {
-                        break;
+
+                // At this point we have found something to unescape, allocate a string for it.
+                // We know it'll be shrinking by at least '1' backslash.
+                let mut unescaped = String::with_capacity(re_str.len() - 1);
```

Comment:
I think `String::new()` is *probably* better here since we don't know by how much less than `re_str.len()-1` we actually need. I can be convinced otherwise though.

Reply:
Easy enough to change back if we can find some evidence that avoiding resizing is, beneficial.

Should fixed in https://github.com/softdevteam/grmtools/pull/333/commits/819f4e9b626ae3e7f5d631757611c349a0013953